src/sources.cpp

```diff
@@ -436,4 +436,211 @@ void fields::add_volume_source(component c, const src_time &src, const volume &w
+  };
+  add_volume_source(c, src, where, A, amp);
+}
+
+static gaussianbeam *global_gaussianbeam_object = 0;
+static component global_gaussianbeam_component;
+
+static std::complex<double> gaussianbeam_ampfunc(const vec &p) {
+  std::complex<double> EH[6];
+  global_gaussianbeam_object->get_fields(EH, p);
+  switch (global_gaussianbeam_component) {
+  case Ex: return EH[0];
+  case Ey: return EH[1];
+  case Ez: return EH[2];
+  case Hx: return EH[3];
+  case Hy: return EH[4];
+  case Hz: return EH[5];
+  default: abort("invalid component in gaussianbeam_ampfunc");
+  }
+}
+
+static int gaussianbeam_has_tm() {
+  return abs(global_gaussianbeam_object->get_E0(2)) > 0;
+}
+
+static int gaussianbeam_has_te() {
+  return abs(global_gaussianbeam_object->get_E0(0)) > 0 || abs(global_gaussianbeam_object->get_E0(1)) > 0;
+}
+
+void fields::add_volume_source(const src_time &src, const volume &where, gaussianbeam beam) {
+  global_gaussianbeam_object = &beam;
+  direction d = normal_direction(where);
+  component cE[3] = {Ex, Ey, Ez}, cH[3] = {Hx, Hy, Hz};
+  int n = (d == X ? 0 : (d == Y ? 1 : 2));
+  if (d == NO_DIRECTION) {
+    n = where.in_direction(X) == 0
+            ? 0
+            : where.in_direction(Y) == 0 ? 1 : where.in_direction(Z) == 0 ? 2 : -1;
+    if (n == -1)
+      abort(
+          "can't determine source direction for non-empty source volume with NO_DIRECTION source");
+  }
+  int np1 = (n + 1) % 3;
+  int np2 = (n + 2) % 3;
+  // Kx = -Hy, Ky = Hx   (for d==Z)
+  global_gaussianbeam_component = cH[np1];
+  add_volume_source_check(cE[np2], src, where, gaussianbeam_ampfunc, +1.0, cE[np2], d,
+                          gaussianbeam_has_tm(), gaussianbeam_has_te());
+  global_gaussianbeam_component = cH[np2];
+  add_volume_source_check(cE[np1], src, where, gaussianbeam_ampfunc, -1.0, cE[np1], d,
+                          gaussianbeam_has_tm(), gaussianbeam_has_te());
+  // Nx = +Ey, Ny = -Ex  (for d==Z)
+  global_gaussianbeam_component = cE[np1];
+  add_volume_source_check(cH[np2], src, where, gaussianbeam_ampfunc, -1.0, cH[np2], d,
+                          gaussianbeam_has_tm(), gaussianbeam_has_te());
+  global_gaussianbeam_component = cE[np2];
+  add_volume_source_check(cH[np1], src, where, gaussianbeam_ampfunc, +1.0, cH[np1], d,
+                          gaussianbeam_has_tm(), gaussianbeam_has_te());
+}
+
+gaussianbeam::gaussianbeam(const vec &x0_, const vec &kdir_, double w0_, double freq_,
+                           double eps_, double mu_, std::complex<double> E0_[3]) {
+  if (x0.dim == Dcyl) abort("wrong dimensionality in gaussianbeam");
+
+  x0 = x0_;
+  kdir = kdir_;
+  w0 = w0_;
+  freq = freq_;
+  eps = eps_;
+  mu = mu_;
+  for (int j=0; j<3; ++j)
+    E0[j] = E0_[j];
+}
+
+/* Compute the E/H fields EH[6] (6 components) at point x in 3d.
+
+   Adapted from code by M. T. Homer Reid in his SCUFF-EM package
+   (file scuff-em/src/libs/libIncField/GaussianBeam.cc), which is GPL v2+. */
+
```

Comment:
You should copy and paste the copyright notice from that source file (https://github.com/HomerReid/scuff-em/blob/9c6d0cb7695463af803dee8d04cdae939740cdcc/libs/libIncField/GaussianBeam.cc#L1-L18) verbatim here.

Reply:
also copy and paste the description with the reference (https://github.com/HomerReid/scuff-em/blob/9c6d0cb7695463af803dee8d04cdae939740cdcc/libs/libIncField/GaussianBeam.cc#L20-L43)

and link the URL https://github.com/HomerReid/scuff-em/blob/master/libs/libIncField/GaussianBeam.cc

Reply:
We should also copy and paste the full copyright notice into https://github.com/NanoComp/meep/blob/master/src/near2far.cpp above the `green3d` function, for that matter.

Reply:
Fixed.